### PR TITLE
Adding cartesian communicator methods

### PR DESCRIPTION
**Pull request by csegarragonz**
As suggested in faasm/faasm#351

### CODE REVIEW

File: src/scheduler/MpiWorld.cpp [hunk 1]
```diff
@@ -176,6 +176,54 @@ std::string MpiWorld::getHostForRank(int rank)
...
+    }
+
+    // Compute the corresponding rank in a 2-dim grid given the original process
+    // rank.
+    // Note - this operation, when restricted to 2dims is quivalent to:
+    // coords = {rank / sideLength, rank % sideLength}
+    int sideLength = static_cast<int>(std::floor(std::sqrt(this->size)));
+    int nprocs = sideLength * sideLength;
+    if (rank >= nprocs) {
+        for (uint8_t i = 0; i < MPI_CART_MAX_DIMENSIONS; i++) {
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:193]: Using an 8-bit unsigned integer as an array index and loop variable seems strange, what's the reason for this?

Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:193]: Using an 8-bit unsigned integer as an array index and loop variable isn't wrong, but it's not in the same style as the rest of the code (which would use an int). What's the reason for this?

File: src/scheduler/MpiWorld.cpp [hunk 2]
```diff
@@ -176,6 +176,54 @@ std::string MpiWorld::getHostForRank(int rank)
...
+}
+
+void MpiWorld::getRankFromCoords(int* rank, int* coords)
+{
+    int sideLength = static_cast<int>(std::floor(std::sqrt(this->size)));
+    int prank = 0;
+    int factor = 1;
+
+    int i = MPI_CART_MAX_DIMENSIONS - 1;
+    for (; i >= 0; --i) {
```
Comment (Shillaker) [inline src/scheduler/MpiWorld.cpp:216]: Then the style of this for loop is also different, we usually initialise loop variables in the loop declaration (and this is using an int whereas the one above is using `uint8_t`)

**Review by Shillaker — CHANGES_REQUESTED** [review]
Looks good other than questions around for loop syntax

File: src/scheduler/MpiWorld.cpp (hunk 2 above)
Comment (csegarragonz) [reply src/scheduler/MpiWorld.cpp:216]: Funny enough, I had to change the type in this case (as using an unsigned int would overflow) but didn't bother looking what was used elsewhere.

**Review by Shillaker — APPROVED** [review]

**Pull request merged**